### DISCUSSION TITLE
Page key not being "normalized" any longer

### DISCUSSION THREAD

**Discussion by leuchtdiode** [pimcore/pimcore#9002] [Q&A / Support] [answered]
Hi,

until a few versions ago when someone created a new page, the key in the third input field got normalized (replaced blanks and special chars with dashes, etc.) automatically.

Why was this feature removed?

E.g. blanks are now allowed which leads to very ugly URLs. 

We developers know this, but explaining this to an unexperienced content manager is a bit "painful" and leads to before mentioned ugly URLs with %20 in it.

Is there a way to enable this feature again?

Thx & BR,
Alex

**Comment by NiklasBr** [comment]
We considered that a feature in our project because the admin UI is meant to be read in a different context than URL:s.

We use the Pretty URL field for the actual URL, you could use [Document Events to sanitize/normalize/prettify](https://github.com/pimcore/pimcore/blob/master/lib/Event/DocumentEvents.php) them automatically.

**Reply by leuchtdiode** [reply]
Thanks for your answer. I will try to rewrite the key via your suggested events.

BR,
Alex

**Comment by leuchtdiode** [comment, marked as answer]
I now ended up with my "german letter" solution like this. Maybe it's useful for someone else too:

```
app/config/services.yml

...
AppBundle\EventListener\DocumentListener:
        tags:
            - { name: kernel.event_listener, event: pimcore.document.preUpdate, method: normalizeKey }
            - { name: kernel.event_listener, event: pimcore.document.preAdd, method: normalizeKey }
...
```

The referenced listener looks like this:

```
AppBundle\EventListener\DocumentListener

<?php
namespace AppBundle\EventListener;

use Pimcore\Event\Model\DocumentEvent;
use Pimcore\Event\Model\ElementEventInterface;

class DocumentListener
{
	public function normalizeKey(ElementEventInterface $e)
	{
		if (!$e instanceof DocumentEvent)
		{
			return;
		}

		$doc = $e->getDocument();

		$key = strtolower($doc->getKey());

		$key = str_replace(
			[
				'&',
				'ß',
				'ü',
				'ö',
				'ä',
			],
			[
				'und',
				'ss',
				'ue',
				'oe',
				'ae',
			],
			$key
		);

		// replace non letter or digits by -
		$key = preg_replace('~[^\pL\d]+~u', '-', $key);

		// transliterate
		$key = iconv('utf-8', 'us-ascii//TRANSLIT', $key);

		// remove unwanted characters
		$key = preg_replace('~[^-\w]+~', '', $key);

		// trim
		$key = trim($key, '-');

		// remove duplicate -
		$key = preg_replace('~-+~', '-', $key);

		$doc->setKey($key);
	}
}
```

BR,
Alex
